Add string registers to sbs

perowle opened the pull request:
drivers: fuel_gauge: sbs_gauge: Add support for string registers

The string registers (chemistry, manufacturer name, device name) were not implemented.  Implemented
by requiring application code to provide a buffer where the strings will be copied once they're
retrieved from the fuel gauge.  Adding the full character arrays to the fuel_gauge_get_property union
would have been expensive, in the sense that every instantiation of the structure would require 21 bytes
for the 'value' field, where making the 'value' field a pointer, limits the maximum size to 4 bytes.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

dcpleung committed:
west.yml: update hal_xtensa for intel_adsp_ace15_mtpm

This udpates the manifest SHA to use the latest Xtensa HAL
commit which is to support building intel_adsp_ace15_mtpm
using Zephyr SDK.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>
Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

tejlmand committed:
cmake: generalize VERSION infrastructure for better reuse

Generalize the VERSION and version.h generation so that the same
infrastructure can be reused for generating other version related
header files, such as an application version header.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
cmake: generated application version header when VERSION file is present

Allow applications to create a VERSION file in their application folder.

The presence of a VERSION file will create a app_version_h target which
generates an app_version.h file using same infrastructure as the Zephyr
kernel version.h generation.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

tejlmand committed:
west: sign: Add Kconfig with application version

Adds a Kconfig option which controls the version of the application
to use when the image is signed using imgtool.

When an application VERSION file is present, the default value will be
identical to the application version, else it will be 0.0.0+0, but a
project may still decide another value, if it so prefers.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

Commit:
net: context: Set priority based on DSCP

If a socket has DSCP set then the packets from the socket should also be
marked with appropriate priority in case traffic classes are used in
networking stack.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

SgrrZhf committed:
driver: uart: pl011_sbsa: refine creating device instance

Create pl011_sbsa device instance via the DT_INST_FOREACH_STATUS_OKAY
macro.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
driver: uart: pl011: fix interrupt driven API

API function:
    - `pl011_irq_tx_enable` is expected to enable and trigger TX interrupt.
    Due to HW limiation, PL011 won't trigger TX interrupt if some data
    wasn't filled to TX FIFO at the beginning. So that `isr_cb` must be
    called at first time to enable TX irq.

    - `pl011_irq_tx_ready` will return true when FIFO can accept more
    data. Here we don't need wait TX FIFO to be empty.

    - `pl011_irq_tx_complete` will return true when all data have been
    sent from the shift register.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

dcpleung committed:
soc: intel_adsp: update toolchain for cavs25

This updates the toolchain used for intel_adsp_cavs25 to
the proper toolchain.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
soc: intel_adsp/ace: update SOC_TOOLCHAIN_NAME

Update the SOC_TOOLCHAIN_NAME to intel_ace15_mtpm so that
we use the correct overlay in Xtensa HAL module. Note that
ace20_lnl will also be using this as well. That will change
once we have a proper toolchain for ace20_lnl.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

PerMac committed:
tests: samples: Resolve duplicates in names.

Several duplicates were found with
scripts/twister -T samples/ -T tests/ --dry-run --list-test-duplicates
This is an issue since duplicated names causes overwriting of
results. Most duplicates looked like obvious copy-pase errors.
New names where addopted looking at other tests in the same yaml
or looking at the directory/descriptio.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

kartben committed:
mailmap: fixed (all?) author names inconsistencies

Ran Levensthein diff on all author names to spot and fix
inconsistencies.
Used authors' most recently used name and emails as canonical.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

gautierg-st committed:
drivers: adc: rewrite stm32 adc enable

Some STM32 families that need to check the ADRDY flag after enabling the
ADC were not doing it. The #ifdef has been updated to fix that.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

MulinChao committed:
intc: miwu: npcx: improve interrupt latency of miwu input events

To reduce the interrupt latency of MIWU events, the driver prepares a
dedicated callback function item list for each MIWU group in this PR. We
needn't check the MIWU table and group of the event in ISR. And the
maximum item number of each list is also limited to 8. After applying
this PR, the interrupt latency reduces to ~10us consistently.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

gautierg-st committed:
drivers: adc: clean stm32 adc calibration

Rework and clean some code around STM32 ADC calibration

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

Commit:
drivers: fuel_gauge: sbs_gauge: Add support for block registers

The block registers (chemistry, manufacturer name, device name) were
not implemented. Implemented by adding a new api interface for
retrieving block properties. fuel_gauge_get_block_property has been
added, and uses a memory buffer allocated in the app in order to
store the fuel gauge information.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

Commit:
drivers: fuel_gauge: sbs_gauge: Add support for block registers

The block registers (chemistry, manufacturer name, device name) were
not implemented. Implemented by adding a new api interface for
retrieving block properties. fuel_gauge_get_block_property has been
added, and uses a memory buffer allocated in the app in order to
store the fuel gauge information.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

Commit:
drivers: fuel_gauge: sbs_gauge: Add support for block registers

The block registers (chemistry, manufacturer name, device name) were
not implemented. Implemented by adding a new api interface for
retrieving block properties. fuel_gauge_get_block_property has been
added, and uses a memory buffer allocated in the app in order to
store the fuel gauge information.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

Commit:
drivers: fuel_gauge: sbs_gauge: Add support for block registers

The block registers (chemistry, manufacturer name, device name) were
not implemented. Implemented by adding a new api interface for
retrieving block properties. fuel_gauge_get_block_property has been
added, and uses a memory buffer allocated in the app in order to
store the fuel gauge information.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

aaronemassey commented:
Hey thanks for making this issue! @madhurimaparuchuri and I are working on this at the moment and it's basically one of our next immediate steps with the API. We've already started on an approach that'll add another function that takes memory allocated by the client. 

> union
would have been expensive

I agree having them in the union would just make that union way too large.

protterc commented:
> Hey thanks for making this issue! @madhurimaparuchuri and I are working on this at the moment and it's basically one of our next immediate steps with the API. We've already started on an approach that'll add another function that takes memory allocated by the client.
> 
> > union
> > would have been expensive
> 
> I agree having them in the union would just make that union way too large.

Thanks Aaron
Would you prefer I close this PR, and wait for your work?  Being new to open source community, trying to understand the appropriate way to contribute, know what's on the roadmap, and to prevent collisions like this?  
Peter

aaronemassey commented:
> Thanks Aaron Would you prefer I close this PR,: 

Whoops, sorry thought this was GitHub issue probably because I was looking at GitHub too early in the morning.

Let's keep it up! We can always contribute to and or review your PR.

>  Being new to open source community, trying to understand the appropriate way to contribute

No worries, I'm still relatively new to Zephyr GitHub and more used to the ChromiumOS tree. Making a PR like this is totally fine. We can also always chat in the `battery` Discord channel on the Zephyr Discord server.

Making a PR like this is fine IMO. For proposing a significant API change we'd normally ask for folks to file an RFC first. I think since this is iterating minimally enough on the existing API that it doesn't warrant an RFC.

@madhurimaparuchuri and I can review this.

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -156,6 +165,16 @@ struct fuel_gauge_get_property {
+		uint16_t manufacturer_date;
+		/** FUEL_GAUGE_SERIAL_NUMBER */
+		uint16_t serial_number;
+		/** FUEL_GAUGE_MANUFACTURER_NAME */
```
So the direction we wanted to go with the API was adding another function that's solely for larger array-backed properties.

I'm wary to extend the existing API like this because it's potentially easy to use it without setting the pointer and cause hard to debug memory corruption bugs. Sure that'd be the client's fault, but I think that'd be a common enough mistake that I'd like to protect against it.

The direction we were heading was to add a function like so: 

```
__syscall int fuel_gauge_get_big_prop(const struct device *dev, struct fuel_gauge_big_property *p, size_t dst_len, void *dst);
```

or if we forego the whole struct parameter:

How would you feel about this change to the API? Would you be okay changing your PR to reflect a design like that?

@teburd do you have any thoughts here?

protterc replied:
I do check for a non-NULL pointer before using the buffer, but get your point.  Even if someone assigns a buffer to the pointer in the union, there's really no guarantee that they've assigned a large enough buffer, whereas the above call, since you're passing a size, kind of ensures that the buffer should work - or at least takes a little guesswork out of the call.  
Only downside is that the user has to know to call 'get big' for the string properties, but that seems like a reasonable tradeoff.  Let me rework the PR to include this.

protterc replied:
Alternately, what if we just added the size to the fuel_gauge_get_property structure?  Along with checking the pointer, can check that the size is non-zero.  The likelyhood of trouble using the function should be minimized in this way, and string properties could still be retrieved using the same API as other registers.

Still would have the opportunity to specify a buffer, and the wrong size, but even with the new API, could do that, and cause memory corruption issues.

Structure would look like below:
		struct fuel_gauge_get_property props[] = {
			{
				.property_type = FUEL_GAUGE_VOLTAGE,
			},
			{
				.property_type = FUEL_GAUGE_DEVICE_CHEMISTRY,
				.value.device_chemistry = (char *)chem_buffer,
				.value.buf_len = sizeof(chem_buffer),
			},
		};

aaronemassey replied:
> I  do check for a non-NULL pointer before using the buffer

We're also not guaranteed that the pointer will be NULL if they haven't set it. That's what really scares me as well as the length of the buffer as you mentioned. Memory errors are scary.

> Still would have the opportunity to specify a buffer, and the wrong size, but even with the new API, could do that, and cause memory corruption issues.

Similarly I'm concerned that if we do this a client may forget to set one of these fields. Unfortunately we can't assume these values will be initialized to zero/NULL. If we could assume such initialization, I'd be 100% agreement for what you just described.

@teburd What are your thoughts here? I tend to err on the side of caution.

protterc replied:
@aaronemassey - updated to make the string properties separate from the "regular" registers.  I'm allocating the memory in the union, so should be safe from a memory perspective.  Thoughts?

aaronemassey replied:
Apologies for keeping you waiting. Was sick yesterday.

> should be safe from a memory perspective

This is definitely memory safe.

So I only see two problems with the API as it stands now.
1) It's pretty specific to NULL terminated strings instead of arbitrary data that we do want to support.
2) No matter what string/big property I want, I have to allocate the max size of bytes on my stack. It's the drawback of a union.

`1.` may be solved easy enough by adding a length field to our new struct here that gets set by the driver and read by the client.

`2.` is a bit trickier. For the properties we have now, I don't think we care about allocating 21 bytes even if we're just grabbing 10 bytes or so. However, there may come a day that we have a 64 byte property to support. We certainly have up to 32 byte properties to support already for SBS. I'd rather not require a client to allocate 32 or 64 bytes on the stack per struct allocated regardless if they're just getting a 10 byte value.

How would you feel if we went with the approach I mentioned above:

```
__syscall int fuel_gauge_get_big_prop(const struct device *dev, struct fuel_gauge_big_property *p, size_t dst_len, void *dst);

```

Where we let clients decide how much memory to give us with the `dst_len` and `dst` parameters which may be informed by constants defined in the `fuel_gauge.h` header like `SBS_CHEMISTRY_MAX_LEN`.

However, this version I just described has one obvious drawback: We have to make multiple calls to get each parameter. This can get significant if syscalls are enabled because we're potentially making multiple syscalls. But I don't think that's a huge price to pay given most of these string/array-backed properties are typically retrieved only once and then forever cached. I also think since these tend to be larger properties with many multiple bytes to read, there's less gain from optimizing burst reads on the driver level like we discussed on Discord; let me know if you think that's not the case.

What do you think @protterc ?

protterc replied:
Hope you're feeling better.  

Good point about the stack, it could get quite large for the one call where you're reading string properties, and could cause apps to have to oversize the task stack, just for this one call to get the static string properties.

I'm thinking of changing the name, not to 'big', but to 'block' - that's how they're referenced in the SBS spec, and it seems to fit well.  They do also reference them as strings, but as you pointed out, even though they're not typical null-terminated C strings, we should probably not overload that nomenclature.

So, I'm going to use your API above, where the app will allocate the memory, and pass it down to the driver.  The struct won't have a 'value' field, just the status, etc.  One nice side effect of creating the call this way is that I can use the 'block' call to also read all the dynamic registers in one shot - don't have to add that logic to the driver to decide if they're all sequential.

protterc replied:
Updated to the requested API

aaronemassey commented on drivers/fuel_gauge/sbs_gauge/sbs_gauge.h:
```diff
@@ -11,41 +12,45 @@
-#define SBS_GAUGE_CMD_DEVICE_NAME           0x21 /* DeviceName */
-#define SBS_GAUGE_CMD_DEVICE_CHEM           0x22 /* DeviceChemistry */
-#define SBS_GAUGE_CMD_MANUFACTURER_DATA     0x23 /* ManufacturerData */
+#define SBS_GAUGE_CMD_MANUFACTURER_ACCESS 0x00 /* ManufacturerAccess */
```
Thank you, I sometimes forget to do this and really need to make a git hook.

sreeramIfx committed:
drivers: bluetooth: Add Infineon Bluetooth driver

Add initial version of the Bluetooth driver for
the cy8cproto_063_ble board

Signed-off-by: Sreeram Tatapudi <sreeram.praveen@infineon.com>

madhurimaparuchuri committed:
drivers: sbs_gauge: Add support for Alarm properties

RemainingCapacityAlarm(r/w) and RemainingTimeAlarm(r/w)

Signed-off-by: Madhurima Paruchuri <mparuchuri@google.com>

madhurimaparuchuri committed:
drivers: sbs_gauge: Reformat the changes

Reformat the files touched by code changes using clang

Signed-off-by: Madhurima Paruchuri <mparuchuri@google.com>

Manimaran-A committed:
drivers: crypto: MEC172x crypto driver supporting hash

Implement zephyr crypto driver hash API's using calls to
MEC172x ROM hash API's. Hardware supports zephyr driver
hash modes: SHA-224, 256, 384, and 512. Driver supports
synchronous (blocking) mode at this time.

Signed-off-by: Manimaran A <manimaran.a@microchip.com>

Manimaran-A committed:
samples: boards: mec172xevb_assy6906 Crypto hash API sample

Sample application exercising Zephyr crypto hash API using
MEC172x driver.

Signed-off-by: Manimaran A <manimaran.a@microchip.com>

jfischer-no committed:
drivers: udc: add USB device controller driver skeleton

Add a USB device controller driver skeleton to use as a starting point
for implementing a specific driver.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

gautierg-st committed:
drivers: adc: add comment to describe different adc versions

Add comment to describe different ADC versions

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

jfischer-no committed:
tests: udc: get UDC driver skeleton tested on native_posix_64

Get UDC driver skeleton tested on native_posix_64.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

Manimaran-A committed:
drivers: gpio: Microchip MEC172x GPIO driver glitch fix

A glitch was observed if a GPIO PIN was configured to a
non-default state by ROM and then Zephyr programs the pin
for the same configuration. Root cause is GPIO hardware
implementing two output bits for each pin. The alternate
output bit is in the pin control register and is r/w by
default. The other bit exists in the GPIO parallel ouput
register and is read-only by default. The hardware actually
reflects the pin's output value into both bits. The fix is
to configure the pin with alternate output bit read-write
and the last step is to disable alternate output which
enabled read-write of the parallel bit. GPIO API's can
then use the GPIO parallel out registers. Add logic to
return an error from the GPIO interrupt configure API if
a pin is not configured as an input. Hardware only performs
interrupt detection if the input pad is enabled.
Hardware supports a pin being configured for both input
and output. Applications should add the GPIO_INPUT flag
to all pin configuration requiring interrupt detection.
The interpretation of input and output flags for the
get configuration API appears to be only one of the
flags can be set. Please refer to the GPIO driver tests.
Updated GPIO interrupt configure to clear the input pad
disable bit due to interrupt detection HW is connected
only to input side of pin.

Signed-off-by: Manimaran A <manimaran.a@microchip.com>

Manimaran-A committed:
drivers: pinctrl: Microchip XEC PINCTRL glitch fix

Glitches were observed if a GPIO pin was configured by
ROM to a non-default state and then Zephyr PINCTRL
reconfigured the pin. The fix involves using the correct
PINCTRL YAML output enable and state flags. Reading the
current spin state and reflecting into new pin configuration
if the pin is output and the drive low/high properties are
not present. We also take advantage of GPIO hardware reflecing
the alternate output value in the parallel output bit before
enabling parallel output mode. Interpret boolean flags with
both enable and disable as do not touch if neither flag is
present. We give precedence to enable over disable if both
flags mistakenly appear. Note, PINCTRL always clears the
GPIO control input pad disable bit.

Signed-off-by: Manimaran A <manimaran.a@microchip.com>

gautierg-st committed:
drivers: adc: cleanup stm32 adc driver

Inverts the logic of some #ifdef, replacing lists of newer series by a
list of NOT older series. This makes it shorter, and more future-proof as
future series are more likely to work out of the box, without need to
manually add a new define in each place.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: add ifdef for stm32f37x adc

Add ifdef around some functions that don't exist for STM32F37x (ADC_V2_5)
like it is done for STM32F1x.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mjchen0 committed:
mimxrt595_evk: add i3c

Add i3c to device tree and the clock init to soc.c

Signed-off-by: Mike J. Chen <mjchen@google.com>

mjchen0 committed:
drivers: i3c: mcux: fix issues when only i2c devices are on the bus

Fixes for bug:
https://github.com/zephyrproject-rtos/zephyr/issues/57560

* don't do CCC if no i3c devices in device tree
* don't wait for MCTRLDONE status when issuing stop
* don't do data part of transfer if buf_sz is 0
* don't limit transfers to only i2c devices in the device tree
  so "i2c scan" shell cmd works as expected

Signed-off-by: Mike J. Chen <mjchen@google.com>

aaronemassey reviewed (changes requested):
Marking as `requested changes` to make sure the discussion shows up on your feed :)

luca-fancellu committed:
drivers: eth_smsc91x: Fix compilation error for assert

Fix a compilation error for the ethernet driver smsc91x that
prevents the build with asserts enabled.

Signed-off-by: Luca Fancellu <luca.fancellu@arm.com>

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -235,6 +273,35 @@ static inline int z_impl_fuel_gauge_get_prop(const struct device *dev,
+ *
+ * @return return=0 if successful, return < 0 if getting property failed, return 0 on success
+ */
+__syscall int fuel_gauge_get_block_prop(const struct device *dev,
```
I like this a lot and it also gives the flexibility of getting multiple block properties if we want if we have a length attribute in the block prop struct. We probably want an example of its usage in that way.

aaronemassey reviewed (changes requested):
I think we're quite close here! At least IMO!

protterc commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -156,9 +170,22 @@ struct fuel_gauge_get_property {
+struct fuel_gauge_get_block_property {
+	/** Battery fuel gauge property to get */
+	uint16_t property_type;
+
```
The fuel gauge doesn't return a null terminated string - it returns a block of bytes, with the first character being the actual length, and the rest of the bytes being the message.  So, to create a null terminated string, the application would need to do something like:
		sprops.property_type = FUEL_GAUGE_MANUFACTURER_NAME;
		ret = fuel_gauge_get_block_prop(dev, &sprops, SBS_GAUGE_MANUFACTURER_NAME_MAX_SIZE, buffer);
		if (ret < 0) {
			printk("Error: cannot get properties\n");
		} else {
			if (ret != 0) {
				printk("Warning: get block property failed\n");
			}

			if (sprops.status == 0) {
				**buffer[(int)buffer[0]+1] = 0;**
				printk("Mfg name %s\n", &buffer[1]);
			}
		}

That said, I was thinking of actually adding a length field, not specifically to tell the user the actual length, but rather to limit/protect the user.  For instance, if the user requests a read of 1000 bytes, compare against the property type, and take the MIN(userlen, propertyTypeMAXLENGTH) as the return value.

I guess another check the driver could do is to look at that first byte, and if it's greater than the above min - say the user only gives a buf_len of 3, but the first byte is 10, should probably return a -EFAULT error or something.

Then the question would be - should the driver read the first byte and set return length, and then shift the byte array, so that the length is no longer encoded in the return?  On the one hand, it makes it easier for the user.  On the other hand, the user is probably expecting the raw string back from the fuel gauge, so the driver is kind of over-stepping it's boundaries.

aaronemassey replied:
@protterc 

Taking a step back, we might want to solve this by bringing the union back.

What if we go back to using a union with array pointers of different types that get defined by our API? So a use case would look something like:

```
fuel_gauge_get_block_prop(dev, &sprops, SBS_GAUGE_MANUFACTURER_NAME_MAX_SIZE, buffer);

struct abstract_type *foo = sprops[0].value.field.
```

Here the driver would just set sprops[0].value to the buffer address. This would allow the API to abstract away the data as something the client can more easily reason with and would be less IC-specific.

So here we'd be letting the client allocate the memory for us but we get to use our IC-specific knowledge on how to get it and then just give a presentation to the client. It also increases the similarity between the two APIs.

Let me know if you want me to provide a more fully fleshed out example.

protterc replied:
Maybe a more complete explanation would help.  Not sure I understand quite what you're thinking.

aaronemassey replied:
@protterc 

Here's a more complete example of what I was thinking. I'd like if drivers can expose arbitrary data to the client and present it as whatever type we feel is abstract and appropriate for them.

```c
/*****************************************************************************/
/*                  Goal: Expose block properties to client                  */
/*****************************************************************************/

#define FUEL_GAUGE_SB_MANUF_NAME 0x1234
#define FUEL_GAUGE_SOME_PROP	 0x1235

/* We could just expose this as a NULL terminated string too */
struct sb_string {
	/** Number of characters in contents */
	uint8_t size;
	/** Contents of sb string minus the size byte */
	char contents[31];
};

struct some_other_type {
	uint32_t foo;
	uint16_t bar;
	uint16_t zeeb;
};

struct fuel_gauge_get_block_property {
	/** Battery fuel gauge property to get */
	uint16_t property_type;
	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
	int status;
	/** Pointer to address in user allocated memory; value is set by driver */
	union {
		/** Raw pointer that is useful for custom types */
		void *raw;
		struct sb_string *manufacturer_name;
		struct some_other_type *some_property;
	} value;
};

/*
 * Function prototype that allows multi-property fetch with one call. We don't need to do this but
 * we can.
 */
int fuel_gauge_get_block_prop(const struct device *dev, struct fuel_gauge_get_block_property *props,
			      size_t props_len, size_t dst_len, void *dst);

/* API Defined above with example usage below */
ZTEST_USER_F(sbs_gauge_new_api, test_block_prop_api_demo)
{

	/* Client use-case with anonumous struct *****************************/
	struct {
		struct sb_string my_manufacturer_name;
		struct some_other_type my_some_property;
	} my_prop_buffer;

	struct fuel_gauge_get_block_property props[] = {
		{
			.property_type = FUEL_GAUGE_SB_MANUF_NAME,
		},
		{
			.property_type = FUEL_GAUGE_SOME_PROP,
		},
	};

	fuel_gauge_get_block_prop(fixture->dev, props, ARRAY_SIZE(props), sizeof(my_prop_buffer),
				  &my_prop_buffer);

	/* These have to be address equal */
	zassert_equal_ptr(props[0].value.manufacturer_name, &my_prop_buffer.my_manufacturer_name);
	zassert_equal_ptr(props[1].value.some_property, &my_prop_buffer.my_some_property);

	/* Client use-case with allocated buffer *****************************/
	uint8_t some_max_bytes = 64;
	uint8_t buffer[some_max_bytes];

	struct fuel_gauge_get_block_property props2[] = {
		{
			.property_type = FUEL_GAUGE_SB_MANUF_NAME,
		},
		{
			.property_type = FUEL_GAUGE_SOME_PROP,
		},
	};

	fuel_gauge_get_block_prop(fixture->dev, props2, ARRAY_SIZE(props2), sizeof(buffer), buffer);

	struct sb_string *my_manuf_name2 = props2[0].value.manufacturer_name;
	struct some_other_type *my_some_property2 = props2[1].value.some_property;

	/* These have to be address equal */
	zassert_equal_ptr(my_manuf_name2, buffer);
	zassert_equal_ptr(my_some_property2, +sizeof(*my_manuf_name2));

	/*
	 * Note above that by using the pointer from the API, the client doesn't have to keep the
	 * properties and anonymous struct fields in order
	 */
}
```

protterc replied:
I'm have a few concerns around this interface.

To clarify my understanding, are the structs sb_string and some_other_type defined in the fuel_gauge.h file?  If so, I'm less worried about size, but if these are user-defined structs, then we could still have potential for memory corruption.

If they are created in fuel_gauge.h, I'm mixed.  On the one hand, they do a great job of defining what's coming back, but on the other hand, they really belong to sbs_gauge.h - they're not generic, but very specific to a given implementation (SBS), so having them in the general interface seems broken.

Having an array of props, but only a single buffer seems problematic in the sense that you need to make sure you order the elements in my_prop_buffer in the same order as the props array.  Also concerned that if there are size problems - like you forget to add a struct, but request a property - do we return an error for the whole batch of properties.  The first few might be OK, because they fit into the mixed buffer, but perhaps are not in the right memory location, depending on where the needed struct wasn't included.

Additionally, you've re-created, in a different way, the issue with consuming potentially a large amount of stack space required of you request all the properties at once.

I guess, given that these are generally static values, and only need to be retrieved once, and cached in the MCU firmware, are we overthinking, or should we just have a general block read function that takes a register, a buffer, and a size, and leave the details of how the data is mapped to the end user, which if they're using these drivers, probably knows what they need, and what they're going to get returned to them based on the data sheet.

aaronemassey replied:
@protterc 

> Let's keep discussing, I'm sure we'll come up with the right interface in the end :-)

Of course! I appreciate the discussion and the time you've taken here along with
the quick responses. Thank you!


> To clarify my understanding, are the structs sb_string and some_other_type defined in the fuel_gauge.h file? 

Yes these were meant to be types defined in `fuel_gauge.h` they were sort of just meant to be rough examples of what could be exposed to a client.

> On the one hand, they do a great job of defining what's coming back, but on the other hand, they really belong to sbs_gauge.h - they're not generic,

Right, I probably should've made it less SBS-specific, but wanted to give a
spec-specific example and a generic example. I think we ought to keep things
generic and not specific to SBS or a particular IC when we reasonably can. We
pay no extra memory cost for adding extra/various struct definitions but we do
pay the maintenance cost of course.

> Having an array of props ...

You're right. Let's keep it to a single prop, it just reduces the complexity for
now and makes it easier for driver developers. We can always expand the API when
there's sufficient motivation to do so.

> Additionally, you've re-created, in a different way, the issue with consuming potentially a large amount of stack space required of you request all the properties at once.

That is correct if the client fetches all the properties at once, but then
that's really just on the client anyway. Although a client could just allocate
that memory in the .data section instead of the stack. **Regardless**, I think
you're right and that we ought to simplify the API by keeping the block
properties to single fetches for now.

> I guess, given that these are generally static values, and only need to be retrieved once

This is definitely true for properties like `manufacturer name` but might not be
true for other arbitrary data in manufacturer_data or some other IC that isn't
SBS.

> which if they're using these drivers, probably knows what they need

It's my intention to keep as much of the datasheet specifics to the driver
implementation and provide some kind of abstraction to the client. Of course, we
may have some leaky abstraction such as properties with an SBS prefix because
they're specific to SBS. But I think we should do best-effort to enable usage of
the API without familiarity with the underlying IC datasheets.


So how would you feel if we basically just dropped the whole multiple property part?

So something like:

```c
int fuel_gauge_get_block_prop(const struct device *dev, struct
fuel_gauge_get_block_property *props, size_t dst_len, void *dst);
```

protterc replied:
@aaronemassey - I think one property at a time makes sense for the block reads, and the API you suggest should work for any case.  

Adding a sbs specific area to the fuel_gauge.h file seems reasonable.  Basically a place to define all the block structures and sizes seems legit.  It stretches the architecture a little bit, but I think it's probably ok.  If people want to use them, great, if people just want to send a char buffer, that's good too.

Want to get your thoughts on one last thing.  RIght now, we support, by name, just reading the 3 string registers.  Given that the API will only read the size that's passed in, would it be OK to allow other property_type's to be sent to the block read function?  What I'm thinking is that I could send a register at the start of all the dynamic data I want to read, and a size, and get it all in one shot.  Would make it so that I can use block read, instead of having to add the logic to sort out the props array, determine if they're all sequential, and then do a block read.  It all seems safe enough, given the user has to provide the buffer and buffer size, just wanted to run it past you.

Peter

aaronemassey replied:
@protterc

> If people want to use them, great, if people just want to send a char buffer,
> that's good too.

We can always enable something like a raw version of a property too. But I'd
rather keep the IC/implementation specifics away from the client and just within
the driver.

> What I'm thinking is that I could send a register at the start of all the
> dynamic data I want to read, and a size, and get it all in one shot...

So I'd like to stay away from clients having to give a specific register.

> Would make it so that I can use block read, instead of having to add the logic
> to sort out the props array, determine if they're all sequential, and then do
> a block read.

I think we should keep that to the current API as-is. Figuring out what
properties are contiguous or not can be done with nominal extra compute time in
the driver. It'll save considerable more time compared to the bus reads,
especially if it's SMBus.

I spent a bit of time thinking up of a rough idea on how to do this and I think
we can do the contiguous bus reads on the fly and even with more than a single
sequence of contiguous registers in a single function call. Some of this logic
can be in a shared fuel gauge library used by the drivers too. But this probably
it's own discussion :)

mbolivar-nordic committed:
scripts: gen_defines: add some missing comments

We generally try to have comments in this file that show the form of
each generated macro. This is particularly important in the
write_vanilla_props() function, since that is called on every node in
the tree and handles generic macros that are widely applicable.

Various generated macros have been added over time that don't have
the corresponding comments; add these now.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
devicetree: support DT_PROP_LEN() on phandle and string

It will be convenient to treat these respectively as degenerate cases
of 'phandles' and 'string-array'. Add support for this and regression
tests. (There's nothing to do in the case of 'phandle' beyond
documenting the guarantee.)

For the record, the other DT_PROP_LEN() tests for each type are in:

  type            test case              property
  ------------    --------------------   ------------
  array           test_arrays            a
  string-array    test_path_props        compatible
  uint8-array     test_arrays            b
  phandles        test_phandles          phs
  phandle-array   test_phandles          pha-gpios
  phandle         test_phandles          ph

Update docstrings and fix some issues in them.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mniestroj committed:
dts: bindings: input: fix comments in longpress example

Example configuration uses 'A' and 'X' key codes for longpress events.
Described behavior shows correct key codes (30 and 45), however comments
near those key codes were invalid for 'X' key. Fix that.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

mbolivar-nordic committed:
devicetree: better DT_PROP_BY_IDX()/DT_FOREACH_PROP_ELEM() support

Support use of these macros with properties of type phandle and
string by allowing iterating over:

- a phandle as if it were a phandles of length 1, for convenience and
  consistency with our ability to take its length (and getting 1)

- the non-null characters in a string: we exclude the null for
  consistency with the return value of DT_PROP_LEN() on string
  properties, which, like strlen(), does not include the null

With this and a previous patch expanding the usage of DT_PROP_LEN(),
there is now a relationship between being able to take a property's
logical length with DT_PROP_LEN() and being able to iterate over its
logical elements with DT_FOREACH_PROP_ELEM(). Explain this in the
documentation.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
devicetree: tests: better DT_FOREACH_PROP_ELEM coverage

Verify expected results for every permissible argument type, including
with a phandle and a string in an inferred binding from /zephyr,user.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

PerMac committed:
twister: Raise error when duplicates found on default

Make twister to raise error by default whenever duplicated test ids
are found. Remove redundant option to list duplicates.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

kartben committed:
mailmap: sort alphabetically

Sort .mailmap file alphabetically

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

de-nordic committed:
mgmt/MCUmgr/grp/img: Rework image list to support DirectXIP

Image list will only report active flag for active image and pending
flag for other slot, but only if image in that slot has higher
version than currently running image.

Signed-off-by: Dominik Ermel <dominik.ermel@nordicsemi.no>

sylvioalves committed:
CODEOWNERS: update ESP32 source files ownership

glaubermaroto is no longer maintaining this sources.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

gmarull committed:
twister: allow loading external configuration

Add a new option to Twister that allows to load alternative
configuration files, `--alt-config-root`. This new option takes an
arbitrary number of root folders where alternative configuration files
can be stored. Twister will check if a test configuration file exists in
any of the alternative test configuration folders. For example, given
`$test_root/tests/foo/testcase.yaml`, Twister will use
`$alt_config_root/tests/foo/testcase.yaml` if it exists.

This feature can be useful if an out-of-tree project needs to run
upstream tests in different configurations, potentially not available
upstream (e.g. an out-of-tree board, or Kconfig setting).

Note that overlaying has been discarded because we can't easily delete
fields from the original configuration file, something that in certain
cases could be required. Current approach may lead to some dupplication,
but guarantees full control of the test configuration.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

valkjsaaa committed:
drivers: sensor: lsm6dsl: Fix lsm6dsl gyroscope full range setting

Bug 1: Fix lsm6dsl_gyro_set_fs_raw does not clear FS125 to register when
setting the full range to be other values.

Bug 2: Fix lsm6dsl_gyro_channel_get does not use the current
gyro_sensitivity when getting data from the gyroscope.

Signed-off-by: Jackie Yang <jackie@jackieyang.me>

wojciechslenska committed:
dts: arm: stm32h5: Add aes node

Add hw crypto support in stm32h5 dtsi. Add missing define in driver.

Signed-off-by: Wojciech Slenska <wsl@trackunit.com>

wojciechslenska committed:
boards: arm: stm32h573i_dk: Enable aes

Activate the AES node in the stm32h573i_dk board.

Signed-off-by: Wojciech Slenska <wsl@trackunit.com>

stephanosio committed:
ci: Switch to CI image v0.26.4

This commit updates the CI workflows to use the CI image v0.26.4, in
order to pull in the Zephyr SDK 0.16.1 release.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
ci: Use Zephyr SDK 0.16.1

This commit updates the CI workflows to use the Zephyr SDK 0.16.1 for
building and testing Zephyr in the CI.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
doc: Update to Zephyr SDK 0.16.1

This commit updates the Getting Started Guide and the Zephyr SDK
documentation to point to the Zephyr SDK 0.16.1 release.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

gautierg-st committed:
dts: bindings: adc: add two compat for stm32 adc

Add two compat for STM32F4-like (F2, F4, F7 & L1) and STM32F1-like (F1
& F37x) ADCs.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
dts: arm: add new adc compatible to stm32

Add the new ADC compatibles for STM32F1 & F373, and for F2, F4, F7 & L1.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: make use of new stm32 adc compatibles

Several sections of the STM32 ADC driver are #ifdef by a combination of the
same SOC defines that share similar IPs. These are F2/F4/F7/L1, and
F1/F37x.
Each of these combinations is now replaced by a specific compatible, which
makes the code a bit lighter and more succinct.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -37,61 +38,70 @@ extern "C" {
 
 /** Battery current (uA); negative=discharging */
-#define FUEL_GAUGE_CURRENT		    FUEL_GAUGE_AVG_CURRENT + 1
+#define FUEL_GAUGE_CURRENT                  FUEL_GAUGE_AVG_CURRENT + 1
```
Thanks again for running `clang-format`. Can we separate the `clang-format` changes into another commit? Other reviewers will probably want the same.

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -156,9 +166,22 @@ struct fuel_gauge_get_property {
+	uint16_t property_type;
+
+	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
+	int status;
```
I suppose since we're only invoking the function with a single property this field is redundant as made obvious by your implementation in the SBS driver.

However, keeping it here would make migrating to supporting more than a single property much easier. What do you think we ought to do?

aaronemassey replied:
@protterc Friendly ping for your comment here.

aaronemassey replied:
Now that I see your comment. SGTM!

aaronemassey commented on drivers/fuel_gauge/sbs_gauge/emul_sbs_gauge.c:
```diff
@@ -100,6 +103,37 @@ static int emul_sbs_gauge_reg_read(const struct emul *target, int reg, int *val)
+
+	switch (reg) {
+	case SBS_GAUGE_CMD_MANUFACTURER_NAME:
+		mfg_name->manufacturer_name_length = 4;
```
Let's use `sizeof(mfg) - 1` and so forth in case someone changes it later and subsequently forgets to change these numbers.

I know I've done something like that before.

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -236,6 +292,35 @@ static inline int z_impl_fuel_gauge_get_prop(const struct device *dev,
+	const struct fuel_gauge_driver_api *api = (const struct fuel_gauge_driver_api *)dev->api;
+
+	if (api->get_block_property == NULL) {
+		return -EFAULT;
```
`return -ENOSYS;` for unsupported

aaronemassey reviewed (changes requested):
Cool! Just some minor comments. I think we're close here as far as I'm concerned.

aaronemassey reviewed (changes requested):
Cool! Just some minor comments. I think we're close here as far as I'm concerned.

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -37,61 +38,70 @@ extern "C" {
+#define FUEL_GAUGE_DEVICE_CHEMISTRY         FUEL_GAUGE_DEVICE_NAME + 1
 /** Reserved to demark end of common fuel gauge properties */
-#define FUEL_GAUGE_COMMON_COUNT FUEL_GAUGE_DESIGN_VOLTAGE + 1
+#define FUEL_GAUGE_COMMON_COUNT             FUEL_GAUGE_DEVICE_CHEMISTRY + 1
```
How do you feel about us adding a separate range of numbers just for these types of properties? That way when a client uses the wrong kind with the wrong API, it's really easy for the driver to tell the client.

aaronemassey replied:
@protterc Friendly ping for your comment here.

aaronemassey replied:
@protterc 
> I like the idea. How about, since it will also affect the register properties, I create a different PR to address - leaving this one a little cleaner? 

Different PR sounds good to me!

> What's your thoughts on COUNT?

Let's make it a MAX with plenty of room for custom properties for downstream users.

> While we're on the subject, any reason these can't/shouldn't be an enum? Something like (would add the comments into the actual enum):
See conversation at https://github.com/zephyrproject-rtos/zephyr/pull/56666#discussion_r1189580654

We had our reasons for this representation (see the linked comment thread) but missed a workaround using `typedef` (I generally avoid `typedef`s). I was going to wait for you to merge this PR before converting things to an `enum` so you wouldn't have to deal with that merge conflict. You're welcome to do the conversion yourself in a different PR too, but I don't mind taking that one since I missed using an `enum` in the first place :)

mmahadevan108 committed:
pm: rt5xx: Enable OS Timer as wakeup source

1. Enable os_timer as a wakeup-source in the board
   dts file.
2. Enable PM_DEVICE when PM is enabled.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
pm: rt6xx: Enable OS Timer as wakeup source

Enable os_timer as a wakeup-source in the board
dts file.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

stephanosio committed:
boards: qemu_x86_tiny: Place Newlib-nano into pinned sections

This commit updates the qemu_x86_tiny linker script to place the
Newlib-nano text and rodata sections into the corresponding pinned
sections so that they are accessible.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

cvinayak committed:
Bluetooth: Controller: Fix CIS offset_min for dissimilar interval

Fix CIS offset_min calculation to consider that ACL interval
and ISO interval can be dissimilar and the offset value has
to be compensated for the latency until the instant at which
the offset is used.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix Central CIS SN/NESN for skipped events

Fix Central CIS SN and NESN for skipped events during CIS
setup. It is possible that CIG event may overlap the ACL or
other events at the CIS create instant, and hence have non
zero lazy value during the CIS setup events until CIS is
established.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

qipengzha committed:
kernel: pin _kernel variable in case of paging

Exception handler(arch/x86/core/ia32/excstub.S) may access
_kernel variable, it will lead to failure when enabled paging,
so make this critical variable pinned.

Signed-off-by: Qipeng Zha <qipeng.zha@intel.com>

mniestroj committed:
boards: arm: xiao_ble: add UF2 runner

Include UF2 runner for the XIAO BLE to allow basic flashing over UF2
without any additional hardware.

Tested with both `xiao_ble` and `xiao_ble_sense` boards.

Signed-off-by: Marcin Niestroj <m.niestroj@emb.dev>

cvinayak committed:
Bluetooth: Controller: Fix sn on Peripheral CIS event abort

Fix SN value on Peripheral CIS event abort. Adjust the SN
value based on the burst number such that the ISO data where
transmitted as normal.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

gautierg-st committed:
dts: arm: st: h7: remove adc3 for stm32h7ax

STM32H7Ax/H7Bx have only two ADCs, so we delete the ADC3 node.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

nordic-krch committed:
logging: Fix user space crash when runtime filtering is on

Logging module data (including filters) are not accessible by
the user space. Macro for creating logs where creating local
variable with filters before checking is we are in the user
context. It was not used in that case but creating variable
was violating access writes that resulted in failure.

Removing variable creation and using filters directly in the
if clause but after checking condition that it is not the
user context. With this approach data is accessed only in
the kernel mode.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

tmon-nordic committed:
usb: device_next: accept padded Mass Storage CDBs

Universal Serial Bus Mass Storage Specification For Bootability requires
that device shall accept Bulk-Only-Transport bCBWCBLength of 12 for all
commands documented in the specification. Allow padding for commands
documented in Bootability specification, but leave others intact.

This fixes Request Sense on Windows 11 which previously failed due to
bCBWCBLength being 12 instead of 6.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

tmon-nordic committed:
usb: device_next: report medium error on Mass Storage failed writes

End Write command with Medium Error / Write Error when underlying disk
write or sync fails.

Signed-off-by: Tomasz Moń <tomasz.mon@nordicsemi.no>

Commit:
net: wifi: Add an event for disconnect complete

This is needed for applications that rely on WPA supplicant being
in disconnected state before issue subsequent commands (e.g., issue scan
immediately after disconnect fails, until disconnect is completed).

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

peter-mitsis committed:
soc: xtensa: Lower HW_STATE_CHECK_DELAY

The macro HW_STATE_CHECK_DELAY represents the number of usec to
busy-wait when waiting for a h/w register state change. This value
has been lowered by 75% to correct a couple of issues related
to Power Management (PM).

1. This resolves a kernel.timer.tickless test failure (see
tests/kernel/timer/timer_api) on the LNL simulator where the test
was taking too long to wake after sleeping for 50 ticks. This
was tracked down to the xtensa SOC power management code where
it was performing a busy wait of 256 usec, which was equivalent
to 2.56 ticks.

2. Furthermore, this delay of 256 usec contradicted the information
in the DTSI (see power-states for d0i3) which states that the
"exit-latency-us" is 100 usec. Reducing this value to 64 helps to
keep that in line.

Signed-off-by: Peter Mitsis <peter.mitsis@intel.com>

finikorg committed:
bluetooth: shell: Add missing return

Avoid invalid bit shift by adding return.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

mmahadevan108 committed:
drivers: uart_mcux: Do not enable and disable the transmitter

Do not enable and disable the UART transmitter

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

marc-hb committed:
soc: xtensa: nxp: invoke `west sign` at `west build` time

This aligns `soc/xtensa/nxp_adsp/` with commit
fad2da39aaaf ("intel_adsp: move `west sign` from `west flash` to earlier
`west build`")

The --if-tool-available option preserves backwards-compatibility:
nothing happens if rimage is not found.

Signed-off-by: Marc Herbert <marc.herbert@intel.com>

Commit:
intel_adsp: ace: Fix heap in the linker

The end of the heap should be the same as _heap_sentry. The current end
marker just covers the range of memory that is explicitly put in
.heap_memand not account until the end of L2_SRAM_BASE +
L2_SRAM_SIZE memory.

Signed-off-by: Flavio Ceolin <flavio.ceolin@intel.com>

wojciechslenska committed:
net: net_pkt: Fix cpp compatibility

Assignment uint8_t to enum without casting, makes error in
cpp compiler.

Signed-off-by: Wojciech Slenska <wsl@trackunit.com>

Commit:
Board: nsim_hs: fix nsim dccm size property

Actual DCCM size is 100000 in nsim_hs_mpuv6, nsim_hs
and nsim_hs3x_hostlink, but their nsim property is
40000. Now fixed the difference to avoid data access
to unpopulated region during picolibc malloc heap init.
In minimal libc, it is set not using malloc as default.

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

FRASTM committed:
tests: drivers: uart async testing on the stm32f103 nucleo

Configure the overlay to have DMA transfer with high priority
when running the tests/drivers/uart/uart_async_api on nucleo_f103rb.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

kartben committed:
usb-c: Remove duplicate typedefs

Remove duplicate declaration for policy_cb_get_src_rp_t and
policy_cb_src_en_t typedefs.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: fix STM32_CLOCK bit field documentation

Fix Doxygen documentation by moving @brief just before STM32_CLOCK
definition.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: add missing param in DT_INST_PIO_PIN_BY_NAME

Add missing '@param name' for DT_INST_PIO_PIN_BY_NAME macro.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: fix Doxygen doc for LOG2(x)

Param x was documented twice.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: fix NUMICRO_PINMUX bit field documentation

Fix Doxygen documentation by moving @brief just before NUMICRO_PINMUX
definition.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

sjanc committed:
bluetooth: tester: Bump supported GATT Prepare Writes to 12

This is required by GATT/SR/GAW/BV-14-C where PTS does a lot of small
writes to characteristic of 512 bytes.

Signed-off-by: Szymon Janc <szymon.janc@codecoup.pl>

Commit:
net: wifi: Fix typo in event

Wrong macro was copied, this causes the event to fail.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

fabiobaltieri committed:
doc: input: expand the kscan compatibility description

Add few details about the kscan compatibility driver, mention the
relevant Kconfig options and an example (which is also present in the
binding file).

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
ci: do_not_merge: check for TSC tag as well

Check for TSC tag in addition to the DNM one, so that PR scheduled for
TSC discussion can't be merged by mistake.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

jfischer-no committed:
drivers: spi_nrfx_spim: bring back get_nrf_spim_frequency

Commit 246393e83044
("drivers: spi: spi_nrfx_spim: Remove nrf_frequency_t handling")'
introduced two changes, one of them is removing the function
get_nrf_spim_frequency with a strange justification.
This change  breaks support for peripherals written in a common way,
where the maximum frequency is set to the maximum supported
by the peripheral, not the controller, see shields for example.

On the occasion of bringing it back, the original function was
refactored to be easier to read and understand.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

nordicjm committed:
cmake: kconfig: Configure KCONFIG_ROOT variable

This configures the KCONFIG_ROOT variable to perform variable
substitution, this allows external applications/modules to supply
variables for this when will then be correctly used.

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

aaronemassey reviewed (changes requested):
Could you respond to a few of the comments I've left?

aaronemassey commented on include/zephyr/drivers/fuel_gauge.h:
```diff
@@ -236,6 +292,35 @@ static inline int z_impl_fuel_gauge_get_prop(const struct device *dev,
+ *
+ * @return return=0 if successful, return < 0 if getting property failed, return 0 on success
+ */
+__syscall int fuel_gauge_get_block_prop(const struct device *dev,
```
I do prefer "buffer" since we're trying to not explicitly tie ourselves to SBS. There are lot of ICs that don't implement SBS but will have similar properties.

Commit:
drivers: fuel_gauge: sbs_gauge: Add support for block registers

The block registers (chemistry, manufacturer name, device name) were
not implemented. Implemented by adding a new api interface for
retrieving block properties. fuel_gauge_get_block_property has been
added, and uses a memory buffer allocated in the app in order to
store the fuel gauge information.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

Commit:
drivers: fuel_gauge: sbs_gauge: Add support for block registers

The block registers (chemistry, manufacturer name, device name) were
not implemented. Implemented by adding a new api interface for
retrieving block properties. fuel_gauge_get_block_property has been
added, and uses a memory buffer allocated in the app in order to
store the fuel gauge information.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

Commit:
drivers: fuel_gauge: sbs_gauge: Add support for block registers

The block registers (chemistry, manufacturer name, device name) were
not implemented. Implemented by adding a new api interface for
retrieving block properties. fuel_gauge_get_block_property has been
added, and uses a memory buffer allocated in the app in order to
store the fuel gauge information.

Signed-off-by: Peter Rowley <perowle@microsoft.com>

Commit:
Merge branch 'add_string_registers_to_sbs' of https://github.com/protterc/zephyr into add_string_registers_to_sbs

protterc commented:
Something went very wrong with rebasing the branch - closing this pull request
